src/openrct2/core/FileIndex.hpp

```diff
@@ -178,32 +179,88 @@ class FileIndex
         return ScanResult(stats, files);
     }
 
-    std::vector<TItem> Build(const ScanResult &scanResult) const
+    void BuildRange(const ScanResult &scanResult,
+                    size_t rangeStart,
+                    size_t rangeEnd,
+                    std::vector<TItem>& items,
+                    std::mutex& printLock) const
     {
-        std::vector<TItem> items;
-        Console::WriteLine("Building %s (%zu items)", _name.c_str(), scanResult.Files.size());
-
-        auto startTime = std::chrono::high_resolution_clock::now();
-        // Start at 1, so that we can reach 100% completion status
-        size_t i = 1;
-        for (auto filePath : scanResult.Files)
+        items.reserve(rangeEnd - rangeStart);
+        for (size_t i = rangeStart; i < rangeEnd; i++)
         {
-            Console::WriteFormat("File %5d of %d, done %3d%%\r", i, scanResult.Files.size(), i * 100 / scanResult.Files.size());
-            i++;
-            log_verbose("FileIndex:Indexing '%s'", filePath.c_str());
+            const auto& filePath = scanResult.Files.at(i);
+
+            if (_log_levels[DIAGNOSTIC_LEVEL_VERBOSE])
+            {
+                std::unique_lock<std::mutex> lock(printLock);
+                log_verbose("FileIndex:Indexing '%s'", filePath.c_str());
+            }
+
             auto item = Create(filePath);
             if (std::get<0>(item))
             {
                 items.push_back(std::get<1>(item));
             }
         }
+    }
+
+    std::vector<TItem> Build(const ScanResult &scanResult) const
+    {
+        std::vector<TItem> allItems;
+        Console::WriteLine("Building %s (%zu items)", _name.c_str(), scanResult.Files.size());
+
+        auto startTime = std::chrono::high_resolution_clock::now();
+
+        const size_t totalCount = scanResult.Files.size();
+        if (totalCount > 0)
+        {
+            JobPool jobPool;
+            std::mutex printLock; // For verbose prints.
 
-        WriteIndexFile(scanResult.Stats, items);
+            std::vector<std::vector<TItem>> containers;
```

Comment:
I think a list of vectors is more desirable here. The vectors don't need to be continguous.

Reply:
lists have generally more overhead due to the nature of each item being a new node so each inserted object is exact one allocation. I'm not sure if its really that big of a difference at the end.

Reply:
I'm talking about the vector of containers, not the vectors of objects (`std::list<std::vector<TItem>>` instead of `std::list<std::list<TItem>>`). The current implementation has as problem that the entire vector could be reallocated when it needs more space which invalidates all references to it. Since you call `resize` before making the references it isn't really a problem now, but even so it'd rather not depend on that.

With a list of containers, you can `push_back` inside the loop that creates the tasks which has almost no overhead over `std::vector`'s `push_back`, and then there is no need to resize.

Reply:
I think vector of vectors is fine. At most you're reallocating the pointers to inner-most data, not the data pointed to. With lists being prohibitively expensive due to their nature in just about any use case, personally I'd implement it as VoV rather than LoV too.

Reply:
I don't think it's fine. This gist shows why: https://gist.github.com/Broxzier/c609de199a421ece580d3f52ff0b7216

The `ref` variable is basically the argument passed to the threads here. If the `containers` vector gets altered after its initialization, we'll get the same problem. These kinds of bugs are annoying to debug, so I'd rather avoid using vectors here than having issues with them later on when someone touches this code. The performance loss is most likely not even noticable.